http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala

```diff
@@ -2,130 +2,222 @@ package org.broadinstitute.dsde.workbench.leonardo
+                val roles = Set("roles/dataproc.worker")
+                if (createCluster) {
+                  googleIamDAO.addIamRoles(googleProject, email, MemberType.ServiceAccount, roles)
+                } else {
+                  googleIamDAO.removeIamRoles(googleProject, email, MemberType.ServiceAccount, roles)
+                }
+              }
+              Future.unit
+            }
           }
+        }
       }
     } getOrElse IO.unit
 
-    // TODO: replace this logic with a group based approach so we don't have to manipulate IAM directly in the image project.
-    // See https://broadworkbench.atlassian.net/browse/IA-1364
-    //
-    // Add the Compute Image User role in the image project to the Google API service account.
-    // This is needed in order to use a custom dataproc VM image.
-    // If a custom image is not being used, this is not necessary.
-    val computeImageUserIO = dataprocConfig.customDataprocImage.flatMap(parseImageProject) match {
-      case None                                                => IO.unit
-      case Some(imageProject) if imageProject == googleProject => IO.unit
-      case Some(imageProject) =>
-        IO.fromFuture(IO(dbRef.inTransaction { _.clusterQuery.countActiveByProject(googleProject) })).flatMap { count =>
-          // Note: don't remove the role if there are existing active clusters in the same project,
-          // because it could potentially break other clusters. We only check this for the 'remove' case,
-          // it's ok to re-add the roles.
-          if (count > 0 && create == false) {
-            IO.unit
-          } else {
-            for {
-              projectNumber <- IO
-                .fromFuture(IO(googleComputeDAO.getProjectNumber(googleProject)))
-                .flatMap(_.fold(IO.raiseError[Long](ClusterIamSetupException(imageProject)))(IO.pure))
-              roles = Set("roles/compute.imageUser")
-
-              // The Dataproc SA is used to retrieve the image. However projects created prior to 2016
-              // don't have a Dataproc SA so they fall back to the API service account. This is documented here:
-              // https://cloud.google.com/dataproc/docs/concepts/iam/iam#service_accounts
-              dataprocSA = WorkbenchEmail(s"service-$projectNumber@dataproc-accounts.iam.gserviceaccount.com")
-              apiSA = WorkbenchEmail(s"$projectNumber@cloudservices.gserviceaccount.com")
-              _ <- retryIam(imageProject, dataprocSA, roles).recoverWith {
-                case e if when400(e) => retryIam(imageProject, apiSA, roles)
+  private def createDataprocImageUserGoogleGroupIfItDoesntExist(
+    dpImageUserGoogleGroupName: String,
+    dpImageUserGoogleGroupEmail: WorkbenchEmail
+  ): IO[Unit] = {
+    logger.info(s"Checking if Dataproc image user Google group '${dpImageUserGoogleGroupEmail}' already exists...")
+    IO.fromFuture[Unit] {
```

Comment:
I'd prefer instead of doing something like `IO.fromFuture(IO{a whole block of code})`, we do 
```
for {
  grouEmailOpt <- IO.fromFuture(IO(googleDirectoryDAO.getGoogleGroup(dpImageUserGoogleGroupEmail)))
 _ <- grouEmailOpt match {
case None =>
???
case Some(x) =>
???
}
} yield ???
```

Reply:
What'd be the benefit of doing that?

Reply:
it defeats the purpose of using `IO` if you're just doing `IO.fromFuture{IO(my whole `Future` based app)}.

also see https://typelevel.org/cats-effect/datatypes/io.html#keep-granularity